### PR TITLE
Improvements to FlatGeobuf DataStore

### PR DESCRIPTION
**Pull request by bjornharrtell**
This improves the community FlatGeobuf DataStore with the following:

* Upgrade to flatgeobuf 3.16.4 (which has no geotools dependency)
* Includes geotools related logic that was previously in flatgeobuf package
* Rename Flatgeobuf to FlatGeobuf
* Support bbox filter via FlatGeobuf index
* Use envelope and feature count from header if available
* Avoid writing empty or header only files
* Support reading larger files
* Handle CRS
* Handle mixed geometry collections
* Can read from any valid URL

This will allow GeoServer flatgeobuf WFS output extension to depend on geotools only.

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- ~~There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.~~
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- ~~Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"~~
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- ~~Documentation has been updated accordingly.~~

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

**Comment by bjornharrtell** [comment]
Hi @jericks I opted to make improvements to your existing module instead.

**Comment by bjornharrtell** [comment]
Argh I think the use of memory mapping is failing on Windows.

**Comment by bjornharrtell** [comment]
All done and green now.

**Comment by bjornharrtell** [comment]
I decided to do another round of work in this PR, list of improvements is updated.

**Comment by bjornharrtell** [comment]
Ok back to all done and green now.

**Comment by bjornharrtell** [comment]
Rebased and squashed on main.

File: modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FeatureConversions.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+     */
+    private static void buildPropertiesVector(
+            SimpleFeature feature, HeaderMeta headerMeta, ByteBuffer target) {
+
+        target.order(ByteOrder.LITTLE_ENDIAN);
+        for (short i = 0; i < headerMeta.columns.size(); i++) {
+            ColumnMeta column = headerMeta.columns.get(i);
+            byte type = column.type;
+            Object value = feature.getAttribute(column.name);
+            if (value == null) continue;
```
Comment (jericks) [inline modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FeatureConversions.java:165]: I would suggest adding brackets around all if else statements for readability.

File: modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatBuffers.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2019, Open Source Geospatial Foundation (OSGeo)
```
Comment (jericks) [inline modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatBuffers.java:5]: If this is new to GeoTools, you should probably update the copyright dates.

File: modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatGeobufDataStore.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+public class FlatGeobufDataStore extends ContentDataStore {
+    final URL url;
+    final File file;
+    final String typeName;
+
+    HeaderMeta headerMeta;
+    SimpleFeatureType phantomFeatureType;
+
+    public FlatGeobufDataStore(URL url) {
```
Comment (jericks) [inline modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatGeobufDataStore.java:50]: Could we also support remote http(s) urls?

**Comment by bjornharrtell** [comment]
Reworked to use InputStream based I/O and verified working in GeoServer with files over 2 GB. Rebased and squashed on main.

File: modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreFactoryTest.java [hunk 4]
```diff
@@ -73,47 +76,51 @@ public void cantProcessWrongName() throws IOException {
...
-        FlatgeobufDataStoreFactory dataStoreFactory = new FlatgeobufDataStoreFactory();
+        FlatGeobufDataStoreFactory dataStoreFactory = new FlatGeobufDataStoreFactory();
         assertFalse(dataStoreFactory.canProcess(params));
     }
 
     @Test
     public void createDataStoreFromFile() throws IOException {
         File file = temporaryFolder.newFile("points.fgb");
         Map<String, Serializable> params = new HashMap<>();
-        params.put("flatgeobuf-file", file);
```
Comment (jericks) [inline modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreFactoryTest.java:84]: I am not sure it matters, but the Shapefile data store also uses "url" as its one and only parameter.  I used something more distinctive so I could tell whether someone was trying to load flatgeobuf just from the Map of params.

**Review by jericks — APPROVED** [review]
Looks really good.  I am excited to see improvements and would like to see this moved to plugin or extension status.

File: modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatGeobufDataStore.java (hunk 3 above)
Comment (bjornharrtell) [reply modules/unsupported/flatgeobuf/src/main/java/org/geotools/data/flatgeobuf/FlatGeobufDataStore.java:49]: good idea, simple now with inputstream - implemented with https://github.com/geotools/geotools/pull/3425/commits/b78c54ac899b2538e815d1cf391bd84ed5b73036

File: modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreFactoryTest.java (hunk 4 above)
Comment (bjornharrtell) [reply modules/unsupported/flatgeobuf/src/test/java/org/geotools/data/flatgeobuf/FlatGeobufDataStoreFactoryTest.java:84]: Agreed might make sense but I don't want to stick out (fx. geojsonstore also have url)

**Comment by bjornharrtell** [comment]
Unfortunately getting inconsistent results on larger files when using url.createStream or if I wrap a FileInputStream with BufferedInputStream.

**Comment by bjornharrtell** [comment]
Problem was that I was using skip that doesn't guarantee actual skip length. Fixed now. :)

**Comment by aaime** [comment]
PR has one conflict (caused by the recent landing of sortpom to reformat the pom files) and has new commits since the last review:

- @bjornharrtell could you resolve the conflict?
- @jericks are you happy with the further changes since last review?

**Comment by bjornharrtell** [comment]
@aaime done.

**Comment by jericks** [comment]
> PR has one conflict (caused by the recent landing of sortpom to reformat the pom files) and has new commits since the last review:
> 
> * @bjornharrtell could you resolve the conflict?
> * @jericks are you happy with the further changes since last review?

Looks good to me!

**Comment by aaime** [comment]
Roger, created the Jira ticket, and squash-merging

**Pull request merged**